openvdb_houdini/abitest/CMakeLists.txt

```diff
@@ -0,0 +1,69 @@
+# Copyright Contributors to the OpenVDB Project
+# SPDX-License-Identifier: MPL-2.0
+#
+#[=======================================================================[
+
+  CMake Configuration for the OpenVDB Houdini ABI Test
+
+  This test verifies ABI compatibility by constructing VDB objects using
+  this version of VDB and then validating using the version of VDB shipped
+  with Houdini and vice-versa. This assumes that the ABI between the two
+  is the same.
+
+#]=======================================================================]
+
+cmake_minimum_required(VERSION 3.3)
+project(OpenVDBHoudiniABITest)
+
+# Monitoring <PackageName>_ROOT variables
+if(POLICY CMP0074)
+  cmake_policy(SET CMP0074 NEW)
+endif()
+
+#########################################################################
+
+message(STATUS "----------------------------------------------------")
+message(STATUS "-------- Configuring OpenVDBHoudiniABITest ---------")
+message(STATUS "----------------------------------------------------")
+
+##########################################################################
+
+# This binary is composed of two object files:
+# - TestABI uses headers from this version of OpenVDB
+# - main uses headers from the version of OpenVDB shipped with Houdini
+
+# Disable some compiler warnings in TBB and Half
+
+add_compile_options(
+  -Wno-conversion
+  -Wno-shorten-64-to-32
+)
+
+# Create an object library that uses the Houdini OpenVDB headers
+
+add_library(TestABI OBJECT TestABI.cc)
+target_include_directories(TestABI PUBLIC ${OPENVDB_INCLUDEDIR})
```

Comment:
as sesis symbols are namespaced, couldn't we instead use them directly without this initial step? For example, to query the abi string:
```
std::string houabi = openvdb::v5_2_sesi::getLibraryAbiVersionString();
std::string abi = openvdb::getLibraryAbiVersionString();
```

etc etc

Reply:
Right now there's no knowledge about which version of OpenVDB is being used for the ABI and Main layers. What you're suggesting sounds like you'd need to parse and extract the namespace from version.h that lives within `OPENVDB_INCLUDEDIR` to pass into the compilation? Wouldn't that be more complicated or am I misunderstanding you here?

Reply:
I'm suggesting you have a sing main and a single source which gets duplicated and is compiled once with our headers and once with Houdini's headers, but the executable is linked against both openvdb and openvdb_sesi. I've quickly tried the following out as an example:

main.cpp
```
#include <iostream>

const char* tmp();
namespace houdini { const char* tmp(); }

int main(int , char* [])
{
    std::cout << tmp() << std::endl;
    std::cout << houdini::tmp() << std::endl;
    return 0;
}
```

tmp/vdb_source.cpp
```
#ifdef HOUDINI
#include </toolkit/include/openvdb/version.h>
namespace houdini {
#else
#include <openvdb/version.h>
#endif

const char* tmp() {
    return openvdb::getLibraryVersionString();
}

#ifdef HOUDINI
}
#endif
```

CMake
```
file(COPY tmp/vdb_source.cpp DESTINATION ${CMAKE_CURRENT_LIST_DIR}/tmp2)

set(SOURCE_FILES
  main.cpp
  tmp/vdb_source.cpp
  tmp2/vdb_source.cpp)

set_property(SOURCE tmp2/vdb_source.cpp APPEND PROPERTY COMPILE_DEFINITIONS "HOUDINI;")

add_executable(sandbox_noddy ${SOURCE_FILES})
target_link_libraries(sandbox_noddy libopenvdb.so)
target_link_libraries(sandbox_noddy libopenvdb_sesi.so)
```

Your solution seems to work so I'm not saying you should do it this way, this is just to provide an alternative

Reply:
Instead of
```
#ifdef HOUDINI
#include </toolkit/include/openvdb/version.h>
namespace houdini {
#else
#include <openvdb/version.h>
#endif
```

You could also set the include path for one of the duplicated sources to point to the houdini toolkit and the other to the vdb installation, resulting simply in:
```
#include <openvdb/version.h>
#ifdef HOUDINI
namespace houdini {
#endif
```

Reply:
Hmm, that doesn't seem too different to what I'm doing here to be honest. The main issue I see with your approach (if I've understood correctly) is that you have now put the declaration(s) of tmp() in one file (main.cpp) and the definition somewhere else (tmp/vdb_source.cpp) and they'd obviously still need to be synced up for this to work - which would likely mean turning to the preprocessor to achieve that.

I think a namespace is probably a little nicer than these extern methods, but I'm not sure there's much in it.

Reply:
@Idclip - I just added some changes to implement what you're suggesting here, it makes the CMake a bit more complicated, but ultimately I think it's better because there's less use of the preprocessor so makes it's all more readable.

Note - I made a few tweaks to your suggestion, rather than copying the file into another destination in the source directory (which we'd have to be careful not to add to the repo), it generates it into the build directory and correctly sets up the dependencies. I think that makes more sense. Let me know what you think.